File: examples/leetcode-out/lua/18/four-sum.lua
function __add(a, b)
	if type(a) == 'table' and type(b) == 'table' then
		local out = {}
		for i = 1, #a do out[#out+1] = a[i] end
		for i = 1, #b do out[#out+1] = b[i] end
		return out
	elseif type(a) == 'string' or type(b) == 'string' then
		return tostring(a) .. tostring(b)
	else
		return a + b
	end
end

function __eq(a, b)
	if type(a) ~= type(b) then return false end
	if type(a) ~= 'table' then return a == b end
	if (a[1] ~= nil or #a > 0) and (b[1] ~= nil or #b > 0) then
		if #a ~= #b then return false end
		for i = 1, #a do if not __eq(a[i], b[i]) then return false end end
		return true
	end
	for k, v in pairs(a) do if not __eq(v, b[k]) then return false end end
	for k, _ in pairs(b) do if a[k] == nil then return false end end
	return true
end

function __index(obj, i)
	if type(obj) == 'string' then
		return __indexString(obj, i)
	elseif type(obj) == 'table' then
		if obj[1] ~= nil or #obj > 0 then
			return obj[(i)+1]
		else
			return obj[i]
		end
	else
		error('cannot index')
	end
end

function __indexString(s, i)
	local len = #s
	if i < 0 then
		i = len + i + 1
	else
		i = i + 1
	end
	if i < 1 or i > len then error('index out of range') end
	return string.sub(s, i, i)
end

function fourSum(nums, target)
	local sorted = (function()
	local items = {}
	for _, n in ipairs(nums) do
		table.insert(items, n)
	end
	table.sort(items, function(a, b)
		local n = a
		local _ka = n
		n = b
		local _kb = n
		return (_ka) < (_kb)
	end)
	local _res = {}
	for _, n in ipairs(items) do
		table.insert(_res, n)
	end
	return _res
end)()
	local n = #sorted
	local result = {}
	for i = 0, (n)-1 do
		if ((i > 0) and __eq(__index(sorted, i), __index(sorted, (i - 1)))) then
			goto __continue0
		end
		for j = __add(i, 1), (n)-1 do
			if ((j > __add(i, 1)) and __eq(__index(sorted, j), __index(sorted, (j - 1)))) then
				goto __continue1
			end
			local left = __add(j, 1)
			local right = (n - 1)
			while (left < right) do
				local sum = __add(__add(__add(__index(sorted, i), __index(sorted, j)), __index(sorted, left)), __index(sorted, right))
				if __eq(sum, target) then
					result = __add(result, {{__index(sorted, i), __index(sorted, j), __index(sorted, left), __index(sorted, right)}})
					left = __add(left, 1)
					right = (right - 1)
					while ((left < right) and __eq(__index(sorted, left), __index(sorted, (left - 1)))) do
						left = __add(left, 1)
						::__continue3::
					end
					while ((left < right) and __eq(__index(sorted, right), __index(sorted, __add(right, 1)))) do
						right = (right - 1)
						::__continue4::
					end
				elseif (sum < target) then
					left = __add(left, 1)
				else
					right = (right - 1)
				end
				::__continue2::
			end
			::__continue1::
		end
		::__continue0::
	end
	return result
end

function test_example_1()
	if not (__eq(fourSum({1, 0, -1, 0, -2, 2}, 0), {{-2, -1, 1, 2}, {-2, 0, 0, 2}, {-1, 0, 0, 1}})) then error('expect failed') end
end

function test_example_2()
	if not (__eq(fourSum({2, 2, 2, 2, 2}, 8), {{2, 2, 2, 2}})) then error('expect failed') end
end

test_example_1()
test_example_2()

File: examples/leetcode-out/php/18/four-sum.php
<?php
function mochi_fourSum($nums, $target) {
	$sorted = (function($tmp){ sort($tmp); return $tmp; })($nums);
	$n = (is_array($sorted) ? count($sorted) : strlen($sorted));
	$result = [];
	for ($i = 0; $i < $n; $i++) {
		if ((($i > 0) && ($sorted[$i] == $sorted[($i - 1)]))) {
			continue;
		}
		for ($j = ((is_array($i) && is_array(1)) ? array_merge($i, 1) : ((is_string($i) || is_string(1)) ? ($i . 1) : ($i + 1))); $j < $n; $j++) {
			if ((($j > ((is_array($i) && is_array(1)) ? array_merge($i, 1) : ((is_string($i) || is_string(1)) ? ($i . 1) : ($i + 1)))) && ($sorted[$j] == $sorted[($j - 1)]))) {
				continue;
			}
			$left = ((is_array($j) && is_array(1)) ? array_merge($j, 1) : ((is_string($j) || is_string(1)) ? ($j . 1) : ($j + 1)));
			$right = ($n - 1);
			while (($left < $right)) {
				$sum = ((is_array(((is_array(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) && is_array($sorted[$left])) ? array_merge(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))), $sorted[$left]) : ((is_string(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) || is_string($sorted[$left])) ? (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) . $sorted[$left]) : (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) + $sorted[$left])))) && is_array($sorted[$right])) ? array_merge(((is_array(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) && is_array($sorted[$left])) ? array_merge(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))), $sorted[$left]) : ((is_string(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) || is_string($sorted[$left])) ? (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) . $sorted[$left]) : (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) + $sorted[$left]))), $sorted[$right]) : ((is_string(((is_array(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) && is_array($sorted[$left])) ? array_merge(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))), $sorted[$left]) : ((is_string(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) || is_string($sorted[$left])) ? (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) . $sorted[$left]) : (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) + $sorted[$left])))) || is_string($sorted[$right])) ? (((is_array(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) && is_array($sorted[$left])) ? array_merge(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))), $sorted[$left]) : ((is_string(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) || is_string($sorted[$left])) ? (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) . $sorted[$left]) : (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) + $sorted[$left]))) . $sorted[$right]) : (((is_array(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) && is_array($sorted[$left])) ? array_merge(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))), $sorted[$left]) : ((is_string(((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j])))) || is_string($sorted[$left])) ? (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) . $sorted[$left]) : (((is_array($sorted[$i]) && is_array($sorted[$j])) ? array_merge($sorted[$i], $sorted[$j]) : ((is_string($sorted[$i]) || is_string($sorted[$j])) ? ($sorted[$i] . $sorted[$j]) : ($sorted[$i] + $sorted[$j]))) + $sorted[$left]))) + $sorted[$right])));
				if (($sum == $target)) {
					$result = ((is_array($result) && is_array([[$sorted[$i], $sorted[$j], $sorted[$left], $sorted[$right]]])) ? array_merge($result, [[$sorted[$i], $sorted[$j], $sorted[$left], $sorted[$right]]]) : ((is_string($result) || is_string([[$sorted[$i], $sorted[$j], $sorted[$left], $sorted[$right]]])) ? ($result . [[$sorted[$i], $sorted[$j], $sorted[$left], $sorted[$right]]]) : ($result + [[$sorted[$i], $sorted[$j], $sorted[$left], $sorted[$right]]])));
					$left = ((is_array($left) && is_array(1)) ? array_merge($left, 1) : ((is_string($left) || is_string(1)) ? ($left . 1) : ($left + 1)));
					$right = ($right - 1);
					while ((($left < $right) && ($sorted[$left] == $sorted[($left - 1)]))) {
						$left = ((is_array($left) && is_array(1)) ? array_merge($left, 1) : ((is_string($left) || is_string(1)) ? ($left . 1) : ($left + 1)));
					}
					while ((($left < $right) && ($sorted[$right] == $sorted[((is_array($right) && is_array(1)) ? array_merge($right, 1) : ((is_string($right) || is_string(1)) ? ($right . 1) : ($right + 1)))]))) {
						$right = ($right - 1);
					}
				} else 
				if (($sum < $target)) {
					$left = ((is_array($left) && is_array(1)) ? array_merge($left, 1) : ((is_string($left) || is_string(1)) ? ($left . 1) : ($left + 1)));
				} else {
					$right = ($right - 1);
				}
			}
		}
	}
	return $result;
}

function mochi_test_example_1() {
	if (!((mochi_fourSum([1, 0, -1, 0, -2, 2], 0) == [[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]))) { throw new Exception('expect failed'); }
}

function mochi_test_example_2() {
	if (!((mochi_fourSum([2, 2, 2, 2, 2], 8) == [[2, 2, 2, 2]]))) { throw new Exception('expect failed'); }
}

mochi_test_example_1();
mochi_test_example_2();
